### PR TITLE
Implement invitation system

### PR DESCRIPTION
**Pull request by kylef**
This includes an invitation system where we can manually add invitations via a command. With an invitation, you can accept or reject it.

This pull request does *NOT* include:

- Email when invitation is created or state is changed
- Lottery command to bass invite people

These will be done in a separate PR.

### CODE REVIEW

File: sotu/templates/rejected.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{% extends "base.html" %}
+
+{% block content %}
+  <div class="entrant row">
+    <p>We're sorry to hear you can't make it {{ entrant.name }}.</p>
```
Comment (segiddins) [inline sotu/templates/rejected.html]: `it,`

File: sotu/models.py [hunk 2]
```diff
@@ -5,7 +7,57 @@
...
+
+
+class Invitation(database.Model):
+    INVITED_STATE = 'invited'  # The user has been invited and has not yet responded
+    ACCEPTED_STATE = 'accepted'  # The user accepted the invitation
+    REJECTED_STATE = 'rejected'  # The user has rejected the invitation
+    ATTENDED_STATE = 'attended'  # The user has checked-in at the event
+
+    STATES = (
+        (INVITED_STATE, 'Invited',),
```
Comment (segiddins) [inline sotu/models.py]: feels like duplicated code

Comment (kylef) [inline sotu/models.py]: Ones a display label and the other is the value.

File: sotu/models.py [hunk 3]
```diff
@@ -5,7 +7,57 @@
...
+    return ''.join(random.choice(characters) for i in range(length))
+
+
 class Entrant(database.Model):
     github_username = peewee.CharField(unique=True)
     name = peewee.CharField()
     email = peewee.CharField(unique=True)
+
+    def invite(self):
+        invitation = Invitation.create(entrant=self, code=generate_random_string())
```
Comment (segiddins) [inline sotu/models.py]: do this in a loop, handling the case where it already exists by generating a new code

File: sotu/static/event.ics [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+LAST-MODIFIED:20150604T203931Z
+CREATED:20150604T203836Z
+X-APPLE-STRUCTURED-LOCATION;VALUE=URI;X-ADDRESS=645 Harrison St\\nSan Fr
+ ancisco CA 94107;X-APPLE-RADIUS=59.89456193118752;X-TITLE=645 Harrison S
+ t:geo:37.783383,-122.395708
+BEGIN:VALARM
+X-WR-ALARMUID:54406726-1EBD-4CA2-8B11-C1525546D18B
+UID:54406726-1EBD-4CA2-8B11-C1525546D18B
+TRIGGER:-PT30M
+DESCRIPTION:This is an event reminder
```
Comment (segiddins) [inline sotu/static/event.ics]: `The CocoaPods State of the Union is starting soon!`

File: sotu/templates/accepted.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{% extends "base.html" %}
+
+{% block content %}
+  <div class="entrant row">
+    <p>See you there {{ entrant.name }}!</p>
```
Comment (segiddins) [inline sotu/templates/accepted.html]: `there, `

File: sotu/templates/invited.html [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{% extends "base.html" %}
+
+{% block content %}
+  <div class="entrant row">
+    <p>Congratulations {{ entrant.name }}! You have an invitation for the State of the Union.</p>
```
Comment (segiddins) [inline sotu/templates/invited.html]: `Congratulations,`

File: sotu/views.py [hunk 7]
```diff
@@ -31,6 +32,52 @@ def get_context_data(self, **kwargs):
         }
 
 
+class InvitationView(JinjaView):
+    template_name = 'invited.html'
+    expected_states = (Invitation.INVITED_STATE,)
```
Comment (segiddins) [inline sotu/views.py]: is this even used anywhere?

Comment (kylef) [inline sotu/views.py]: Yes. see https://github.com/CocoaPods/sotu.cocoapods.org/pull/11/files#diff-39c9d453b82e7398ba8c50707f938682R54

**Comment by segiddins** [comment]
Should probably add links on the entrant view?
Otherwise, looks OK to me, though I'm no expert :)

**Comment by kylef** [comment]
> Should probably add links on the entrant view?

The user won't make it to the entrant view if they have an invite. They instead are sent to `invited.html`

**Pull request merged**